### ISSUE TITLE
Inaccessible base not diagnosed in expression using the pointer-to-member operators

### ISSUE THREAD

**Issue by llvmbot** [llvm/llvm-project#19648]
|  |  |
| --- | --- |
| Bugzilla Link | [19274](https://llvm.org/bz19274) |
| Version | unspecified |
| OS | Linux |
| Attachments | [testcase.cpp](https://user-images.githubusercontent.com/60944935/143749248-dbb51fd3-1917-4ee4-b577-6490855979fa.gz) |
| Reporter | LLVM Bugzilla Contributor |
| CC | @majnemer,@DougGregor |

## Extended Description 

```cpp
struct A { int a = 0; };
struct B : private A { };

int
main (int argc, char *argv[])
{
  int A::* ptr = &A::a;

  B ().*ptr; // ill-formed
}
```

----------------------------------------------------------------

The following should issue a diagnostic saying that our source is ill-formed, `clang` however compiles and accepts it. `A` is not an accessible base of `B`, therefore the line marked `// ill-formed` should error.

`gcc` shows the correct behaviour with the following diagnostic:
```
testcase.cpp: In function ‘int main(int, char**)’:
testcase.cpp:9:9: error: ‘A’ is an inaccessible base of ‘B’
   B ().*ptr; // ill-formed
```

[expr.mptr.oper]p2 states the following:

  > The binary operator `.*` binds its second operand, which shall be
  > of type "pointer to member of T", to its first operand, which
  > shall be of class T or of a class of which T is an unambiguous and
  > accessible base class. The result is an object or a function of
  > the type specified by the second operand.

[ Note: the same applies to `->*` ]

**Comment by zwuis** [comment]
> The following should issue a diagnostic saying that our source is ill-formed, `clang` however compiles and accepts it. `A` is not an accessible base of `B`, therefore the line marked `// ill-formed` should error.

Clang 3.5 rejects this code expectedly.

**Comment by Endilll** [comment]
Both the original example and Richard's example appears to be handled correctly since Clang 3.5:
https://godbolt.org/z/sa8YE9jo7
https://godbolt.org/z/eezr1Yzob

**Comment by llvmbot** [comment]
@llvm/issue-subscribers-clang-frontend

Author: None (llvmbot)

<details>
|  |  |
| --- | --- |
| Bugzilla Link | [19274](https://llvm.org/bz19274) |
| Version | unspecified |
| OS | Linux |
| Attachments | [testcase.cpp](https://user-images.githubusercontent.com/60944935/143749248-dbb51fd3-1917-4ee4-b577-6490855979fa.gz) |
| Reporter | LLVM Bugzilla Contributor |
| CC | @majnemer,@DougGregor |

## Extended Description 

```cpp
struct A { int a = 0; };
struct B : private A { };

int
main (int argc, char *argv[])
{
  int A::* ptr = &amp;A::a;

  B ().*ptr; // ill-formed
}
```

----------------------------------------------------------------

The following should issue a diagnostic saying that our source is ill-formed, `clang` however compiles and accepts it. `A` is not an accessible base of `B`, therefore the line marked `// ill-formed` should error.

`gcc` shows the correct behaviour with the following diagnostic:
```
testcase.cpp: In function ‘int main(int, char**)’:
testcase.cpp:9:9: error: ‘A’ is an inaccessible base of ‘B’
   B ().*ptr; // ill-formed
```


[expr.mptr.oper]p2 states the following:

  &gt; The binary operator `.*` binds its second operand, which shall be
  &gt; of type "pointer to member of T", to its first operand, which
  &gt; shall be of class T or of a class of which T is an unambiguous and
  &gt; accessible base class. The result is an object or a function of
  &gt; the type specified by the second operand.



[ Note: the same applies to `-&gt;*` ]
</details>
